crates/editor/src/element.rs

```diff
@@ -4200,17 +4200,73 @@ impl LineWithInvisibles {
                     line_y,
                 );
 
-            if let Some(allowed_regions) = allowed_invisibles_regions {
-                let invisible_point = DisplayPoint::new(row, token_offset as u32);
-                if !allowed_regions
+            (
+                [token_offset, token_end_offset],
+                Box::new(move |cx: &mut WindowContext| {
+                    invisible_symbol.paint(origin, line_height, cx).log_err();
+                }),
+            )
+        };
+
+        let invisible_iter = self.invisibles.iter().map(extract_whitespace_info);
+        match whitespace_setting {
+            ShowWhitespaceSetting::None => return,
+            ShowWhitespaceSetting::All => invisible_iter.for_each(|(_, paint)| paint(cx)),
+            ShowWhitespaceSetting::Selection => invisible_iter.for_each(|([start, _], paint)| {
+                let invisible_point = DisplayPoint::new(row, start as u32);
+                if !selection_ranges
                     .iter()
                     .any(|region| region.start <= invisible_point && invisible_point < region.end)
                 {
-                    continue;
+                    return;
+                }
+
+                paint(cx);
+            }),
+
+            // For a whitespace to be on a boundary, any of the following conditions need to be met:
+            // - It is a tab
+            // - It is adjacent to an edge (start or end)
+            // - It is adjacent to a whitespace (left or right)
+            ShowWhitespaceSetting::Boundary => {
+                // We'll need to keep track of the last invisible we've seen and then check if we are adjacent to it for some of
+                // the above cases.
+                // Note: We zip in the original `invisibles` to check for tab equality
+                let mut last_seen: Option<(bool, usize, Box<dyn Fn(&mut WindowContext)>)> = None;
```

Comment:
> Box<dyn Fn(&mut WindowContext)>

Did I miss anything or is it always `paint` in this code and we can remove the 3rd tuple item entirely?

Reply:
Line 4253 needs access to the last `paint` lambda in order to draw a previous whitespace if it hadn't been drawn already. This is to catch the fact that since we are scanning from the left, the first whitespace in a boundary section will always be skipped.

I couldn't think of a nicer way to ensure that all characters in a boundary are drawn in one pass without this check on line 4250, but I'm down to change it if the logic seems too convoluted!

Reply:
Oh, that's a lambda capturing `origin`, I see...
Yeah, that all looks like a very good candidate for a refactoring, I'd extract it into a separate method as proposed in the testing approach.